sql: change TestRaceWithBackfill to run against a cluster

vivekmenezes opened the pull request:
The test used to bring up a server and execute a schema change
along with other OLTP operations. It now brings up a cluster and
splits up the test table into many ranges, so that when we transition
the schema change backfill to distsql, this test will also test
OLTP operations in the presence of a distributed execution of the
schema change.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13530)
<!-- Reviewable:end -->

vivekmenezes committed:
distsqlrun: fix race writing to firstErr

vivekmenezes committed:
sql: change TestRaceWithBackfill to run against a cluster

The test used to bring up a server and execute a schema change
along with other OLTP operations. It now brings up a cluster and
splits up the test table into many ranges that reside on different
nodes. This allows testing with a distributed version of the
index/column backfill.

vivekmenezes commented:
@RaduBerinde I've also added a minor fix for a race condition to this PR.

RaduBerinde commented:
CC @andreimatei - I think his PR #13570 will rework error propagation through `MultiplexedRowChannel`

RaduBerinde commented:
LGTM!

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/schema_changer_test.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-KcyZTiAOToz9mrgidz2:-KcyZTiAOToz9mrgidz3:bavpd4p) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/schema_changer_test.go#L36)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/keys"
> 	"github.com/cockroachdb/cockroach/pkg/roachpb"
> 	csql "github.com/cockroachdb/cockroach/pkg/sql"
> ```

nyc, but we should stop using `csql` (in favor of using `gosql` for `database/sql`). I thought I removed all instances a while back but I guess not

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-:-KcyZsr2Hxpvq-aH69It:b-85i648)*
<!-- Sent from Reviewable.io -->

vivekmenezes committed:
sql: eliminate package name csql

vivekmenezes commented:
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/schema_changer_test.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-KcyZTiAOToz9mrgidz2:-KczT6fYUVs6_kDZoL6e:b-s0fgpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/schema_changer_test.go#L36)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

nyc, but we should stop using `csql` (in favor of using `gosql` for `database/sql`). I thought I removed all instances a while back but I guess not
</blockquote></details>

added a commit that eliminate csql from the pkg/sql

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/schema_changer_test.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-KcyZTiAOToz9mrgidz2:-Kd0um3E_TDvbwFuZTAM:be8mby0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/schema_changer_test.go#L36)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

added a commit that eliminate csql from the pkg/sql
</blockquote></details>

Thank you, looks great!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner_test.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-Kd1_7yRX0NusSqpdTjx:-Kd1_7yRX0NusSqpdTjy:b6esdw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/distsql_physical_planner_test.go#L36)):*
> SplitTable splits a range in the table, with each range having a single replica on a certain node

Something doesn't make sense.

---

*[pkg/sql/distsql_physical_planner_test.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-Kd1_GEHw5ChZFNfWv_x:-Kd1_GEHw5ChZFNfWv_y:b-n6iqm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/distsql_physical_planner_test.go#L43)):*
> ```Go
> // TODO(radu): this approach should be generalized into test infrastructure
> // (perhaps by adding functionality to logic tests).
> // TODO(radu): we should verify that the plan is indeed distributed as
> ```

which plan? Comment doesn't make sense here.
The other TODO above should probably go too.

---

*[pkg/sql/distsql_physical_planner_test.go, line 49 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-Kd1a764HF_liFBiIjEm:-Kd1a764HF_liFBiIjEn:b-12ufoa) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/distsql_physical_planner_test.go#L49)):*
> ```Go
> 	tc serverutils.TestClusterInterface,
> 	desc *sqlbase.TableDescriptor,
> 	targetNode int,
> ```

`targetNodeIdx` please. Much confusion in tests around `NodeID`s (which start from 1) vs indexes (0-based).

---

*[pkg/sql/distsql_physical_planner_test.go, line 62 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-Kd1_dHk1agfbz04ucYL:-Kd1_dHk1agfbz04ucYM:b-uymhir) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/distsql_physical_planner_test.go#L62)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	splitKey = rightRange.StartKey.AsRawKey()
> ```

mind adding a comment suggesting why we need to update the `splitKey`? Or rename the first one `candidateSplitKey`.

---

*[pkg/sql/distsql_physical_planner_test.go, line 73 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-Kd1_KtYE7EmYpChtq2B:-Kd1_KtYE7EmYpChtq2C:b-hpsy1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/distsql_physical_planner_test.go#L73)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	if _, err := tc.RemoveReplicas(splitKey, tc.Target(0)); err != nil {
> ```

does this (or even the ` tc.Server(0).SplitRange(splitKey)` above) assume that there was a replica on node idx 0? Mind adding an assertion? Or better yet, make the function more resilient by locating the proper server to perform these commands. I know you didn't write this code, but you're elevating it :)

---

*[pkg/sql/schema_changer_test.go, line 589 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-Kd1_tWdVtNEs3AI38WC:-Kd1_tWdVtNEs3AI38WD:b-oiw6g5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/schema_changer_test.go#L589)):*
> ```Go
> 	// Split the table into multiple ranges.
> 	tableDesc := sqlbase.GetTableDescriptor(kvDB, "t", "test")
> 	// SplitTable moves the right range, so we split things back to front.
> ```

I think this comment (as the similar one from the other test), won't make any sense to a reader tomorrow. What's the "move" got to do with the order in which we need to split?
I'd say either explain more that the range we're splitting needs to be on the first server, or relax that requirement.

---

*[pkg/sql/distsqlrun/base.go, line 166 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-Kd1aghhRwJRv0470e1D:-Kd1aghhRwJRv0470e1E:bv1f3h2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/distsqlrun/base.go#L166)):*
> ```Go
> 	// yet to call Close().
> 	numSenders int32
> 	firstErr   chan error
> ```

just curious, have you seen this pattern used before for "write once" members?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
addressed followup comments on https://github.com/cockroachdb/cockroach/pull/13612

---

Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner_test.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-Kd1_7yRX0NusSqpdTjx:-Kd2VTGbnILLGKQKfxQH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/distsql_physical_planner_test.go#L36)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> SplitTable splits a range in the table, with each range having a single replica on a certain node

Something doesn't make sense.
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner_test.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-Kd1_GEHw5ChZFNfWv_x:-Kd2W6tRLIC-mxbQmKbc:b-ee5jir) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/distsql_physical_planner_test.go#L43)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

which plan? Comment doesn't make sense here.
The other TODO above should probably go too.
</blockquote></details>

reworded both TODOs

---

*[pkg/sql/distsql_physical_planner_test.go, line 49 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-Kd1a764HF_liFBiIjEm:-Kd2WRr3FM8M2PZBdGsj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/distsql_physical_planner_test.go#L49)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`targetNodeIdx` please. Much confusion in tests around `NodeID`s (which start from 1) vs indexes (0-based).
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner_test.go, line 62 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-Kd1_dHk1agfbz04ucYL:-Kd2X3rDcIPDZpeGVx_m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/distsql_physical_planner_test.go#L62)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

mind adding a comment suggesting why we need to update the `splitKey`? Or rename the first one `candidateSplitKey`.
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer_test.go, line 589 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-Kd1_tWdVtNEs3AI38WC:-Kd2kf1aSZV7dv9qCEOF:brlvvty) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/schema_changer_test.go#L589)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think this comment (as the similar one from the other test), won't make any sense to a reader tomorrow. What's the "move" got to do with the order in which we need to split?
I'd say either explain more that the range we're splitting needs to be on the first server, or relax that requirement.
</blockquote></details>

This has nothing to do with the range being on the first server. I've updated the comment

---

*[pkg/sql/distsqlrun/base.go, line 166 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-Kd1aghhRwJRv0470e1D:-Kd2XmUdrKcUAsc4qBMY:bsffkdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db55e8fba4476b9ad2839265f8f61797ee8213c/pkg/sql/distsqlrun/base.go#L166)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

just curious, have you seen this pattern used before for "write once" members?
</blockquote></details>

I don't remember where I saw this, but when I did write it up, I was expecting someone to push back with their own esteemed idea which I wouldn't have a problem incorporating. But then no one did :-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13530#-:-Kd2lf0Ocfo_ngnVejtN:b7o22d0)*
<!-- Sent from Reviewable.io -->
